integrations/dump_restore_test.go

```diff
@@ -0,0 +1,107 @@
+	"code.gitea.io/gitea/modules/util"
+	"code.gitea.io/gitea/services/migrations"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestDumpRestore(t *testing.T) {
+	onGiteaRun(t, func(t *testing.T, u *url.URL) {
+		AllowedDomains := setting.Migrations.AllowedDomains
+		setting.Migrations.AllowedDomains = "loopback"
+		AllowLocalNetworks := setting.Migrations.AllowLocalNetworks
+		setting.Migrations.AllowLocalNetworks = true
+		AppVer := setting.AppVer
+		setting.AppVer = "1.15"
+		defer func() {
+			setting.Migrations.AllowedDomains = AllowedDomains
+			setting.Migrations.AllowLocalNetworks = AllowLocalNetworks
+			setting.AppVer = AppVer
+		}()
+
+		assert.NoError(t, migrations.Init())
+
+		reponame := "repo1"
+
+		basePath, err := os.MkdirTemp("", reponame)
+		assert.NoError(t, err)
+		defer util.RemoveAll(basePath)
+
+		repo := unittest.AssertExistsAndLoadBean(t, &repo_model.Repository{Name: reponame}).(*repo_model.Repository)
+		repoOwner := unittest.AssertExistsAndLoadBean(t, &user_model.User{ID: repo.OwnerID}).(*user_model.User)
+		session := loginUser(t, repoOwner.Name)
+		token := getTokenForLoggedInUser(t, session)
+
+		//
+		// Phase 1: dump repo1 from the Gitea instance to the filesystem
+		//
+
+		ctx := context.Background()
+		var opts = migrations.MigrateOptions{
+			GitServiceType: structs.GiteaService,
+			Issues:         true,
+			Comments:       true,
+			AuthToken:      token,
+			CloneAddr:      repo.CloneLink().HTTPS,
+			RepoName:       reponame,
+		}
+		err = migrations.DumpRepository(ctx, basePath, repoOwner.Name, opts)
+		assert.NoError(t, err)
+
+		//
+		// Verify desired side effects of the dump
+		//
+		d := filepath.Join(basePath, repo.OwnerName, repo.Name)
+		for _, f := range []string{"repo.yml", "topic.yml", "issue.yml"} {
+			assert.FileExists(t, filepath.Join(d, f))
+		}
+
+		//
+		// Phase 2: restore from the filesystem to the Gitea instance in restoredrepo
+		//
+
+		newreponame := "restoredrepo"
+		err = migrations.RestoreRepository(ctx, d, repo.OwnerName, newreponame, []string{"issues", "comments"})
+		assert.NoError(t, err)
+
+		newrepo := unittest.AssertExistsAndLoadBean(t, &repo_model.Repository{Name: newreponame}).(*repo_model.Repository)
+
+		//
+		// Phase 3: dump restoredrepo from the Gitea instance to the filesystem
+		//
+		opts.RepoName = newreponame
+		opts.CloneAddr = newrepo.CloneLink().HTTPS
+		err = migrations.DumpRepository(ctx, basePath, repoOwner.Name, opts)
+		assert.NoError(t, err)
+
+		//
+		// Verify the dump of restoredrepo is the same as the dump of repo1
+		//
+		newd := filepath.Join(basePath, newrepo.OwnerName, newrepo.Name)
+		assert.NoError(t, exec.Command("sed", "-ie", "s/"+reponame+"/"+newreponame+"/g", filepath.Join(d, "repo.yml")).Run())
+		output, err := exec.Command("diff", "-u", filepath.Join(d, "repo.yml"), filepath.Join(newd, "repo.yml")).CombinedOutput()
+		if err != nil {
+			fmt.Println(string(output))
+			assert.NoError(t, err)
+		}
```

Comment:
`sed`: maybe we do not need it, we can just read the file in Go and replace string by code.
`diff`: maybe we can use `assert.Equal`, if the values differ, the framework will output the difference.

Reply:
> sed: maybe we do not need it, we can just read the file in Go and replace string by code.
diff: maybe we can use assert.Equal, if the values differ, the framework will output the difference.

Indeed, I discovered the output of EqualValues is very readable. This was changed as suggested.

---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/30#note_683)